drivers/adc/adc_b91.c

```diff
@@ -0,0 +1,367 @@
+		return -EINVAL;
+	}
+
+	/* Check reference */
+	if (channel_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Selected ADC reference is not supported.");
+		return -EINVAL;
+	}
+
+	/* Check gain */
+	switch (channel_cfg->gain) {
+	case ADC_GAIN_1:
+		pre_scale = ADC_PRESCALE_1;
+		break;
+	case ADC_GAIN_1_4:
+		pre_scale = ADC_PRESCALE_1F4;
+		break;
+	default:
+		LOG_ERR("Selected ADC gain is not supported.");
+		return -EINVAL;
+	}
+
+	/* Check acquisition time */
+	switch (channel_cfg->acquisition_time) {
+	FOR_EACH(CASE_ACK_TIME_TO_HAL_SAMPLE_CYC, (),
+		 3, 6, 9, 12, 15, 18, 21, 24, 27,
+		 30, 33, 36, 39, 42, 45, 48)
+
+	case ADC_ACQ_TIME_DEFAULT:
+		sample_cycl = ADC_SAMPLE_CYC_6;
+		break;
+
+	default:
+		LOG_ERR("Selected ADC acquisition time is not supported.");
+		return -EINVAL;
+	}
+
+	/* Check for valid pins configuration */
+	input_positive = adc_b91_get_pin(channel_cfg->input_positive);
+	input_negative = adc_b91_get_pin(channel_cfg->input_negative);
+	if ((input_positive == (uint8_t)ADC_VBAT || input_negative == (uint8_t)ADC_VBAT) &&
+		channel_cfg->differential) {
+		LOG_ERR("VBAT pin is not available for differential mode.");
+		return -EINVAL;
+	} else if (channel_cfg->differential && (input_negative == (uint8_t)NOINPUTN)) {
+		LOG_ERR("Negative input is not selected.");
+		return -EINVAL;
+	}
+
+	/* Init ADC */
+	m_data.differential = channel_cfg->differential;
+	vref = DT_STRING_TOKEN(DT_DRV_INST(0), vref);
+	sample_freq = DT_STRING_TOKEN(DT_DRV_INST(0), sample_freq);
+	adc_init(vref, pre_scale, sample_freq);
+	adc_set_vbat_divider(ADC_VBAT_DIV_OFF);
+	adc_set_tsample_cycle(sample_cycl);
+
+	/* Init ADC Pins */
+	if (channel_cfg->differential) {
+		/* Differential pins configuration */
+		adc_pin_config(ADC_GPIO_MODE, input_positive);
+		adc_pin_config(ADC_GPIO_MODE, input_negative);
+		adc_set_diff_input(channel_cfg->input_positive, channel_cfg->input_negative);
+	} else if (input_positive == (uint8_t)ADC_VBAT) {
+		/* Single-ended Vbat pin configuration */
+		adc_set_diff_input(ADC_VBAT, GND);
+	} else {
+		/* Single-ended GPIO pin configuration */
+		adc_pin_config(ADC_GPIO_MODE, input_positive);
+		adc_set_diff_input(channel_cfg->input_positive, GND);
+	}
+
+	return 0;
+}
+
+/* API implementation: read */
+static int adc_b91_read(const struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	int status;
+
+	adc_context_lock(&m_data.ctx, false, NULL);
+	status = adc_b91_adc_start_read(dev, sequence);
+	adc_context_release(&m_data.ctx, status);
+
+	return status;
+}
+
+#ifdef CONFIG_ADC_ASYNC
+/* API implementation: read_async */
+static int adc_b91_read_async(const struct device *dev,
+			      const struct adc_sequence *sequence,
+			      struct k_poll_signal *async)
+{
+	int status;
+
+	adc_context_lock(&m_data.ctx, true, async);
+	status = adc_b91_adc_start_read(dev, sequence);
+	adc_context_release(&m_data.ctx, status);
+
+	return status;
+}
+#endif /* CONFIG_ADC_ASYNC */
```

Comment:
this driver seems to support async essentially by spawning a thread that does polling, ie no IRQ/DMA based. this doesn't seem like the right approach to do "async". cc: @teburd

Reply:
This driver does not support IRQs for ADC and because of that it is implemented in the same way as other drivers without IRQ support: adc_ads1x1x.c, adc_emul.c, adc_lmp90xxx.c, adc_mcp320x.c, and adc_npcx.c.

Reply:
cc: @anangl